samples/security/TokenLoginModule.java

```diff
@@ -0,0 +1,250 @@
+                debug("Shared state:");
+                sharedState.forEach((k, v) -> debug(String.format("  %s=%s", k, v)));
+            }
+        }
+    }
+
+    @Override
+    public boolean login() throws LoginException {
+
+        try {
+
+            AuthenticationToken authToken = getTokenFromCallback();
+            if ((authToken == null) || (authToken.isEmpty())) {
+                // Since a token was expected, issue a challenge to the client to provide one.
+                // Calling challenge() will result in a 401 HTTP response back to the client
+                // when an exception is thrown that results in authorization failing. If an
+                // exception is thrown without first calling challenge(), the client will
+                // receive a 403 HTTP response instead.
+                LoginResult loginResult = getLoginResultFromCallback();
+                loginResult.challenge();
+                throw new LoginException("No tokens provided");
+            }
+            debug(String.format("Login: Obtained AuthenticationToken: %s", authToken));
+
+            // Ensure that the token data uses the "Token" authentication scheme.
+            // Note: Compare to "Token" rather than "Application Token" because the scheme
+            // is just the scheme type which is Basic, Negotiate, or Token.
+            String authScheme = authToken.getScheme();
+            if (!authScheme.equals("Token")) {
+                throw new LoginException(
+                    "Expected authentication token scheme 'Token', but received '" + authToken.getScheme() + "'");
+            }
+
+            // The AuthenticationToken contains extra information (such as the scheme), so
+            // extract the token data specifically.
+            String tokenData = authToken.get();
+
+            processToken(tokenData);
+
+            // If we made it here then the token is valid. Save the username into shared state
+            // for downstream login modules to authorize.
+            debug(String.format("Login: Saving username to shared state: %s", username));
+            sharedState.put(USERNAME_KEY, username);
+
+            return true;
+
+        }
+        catch (LoginException e) {
+            debug(String.format("Login: Got a LoginException: %s", e.getMessage()));
+            throw e;
+        }
+        catch (Exception e) {
+            debug(String.format("Login: Got an Exception: %s", e.getMessage()));
+            throw new LoginException(e.getMessage());
+        }
+
+    }
+
+    @Override
+    public boolean commit() throws LoginException {
+
+        // This method associates relevant Principals and Credentials with the Subject.
+        // This login module does not perform authorization, but extracts the username
+        // from a token for downstream login modules to authorize. Therefore there is
+        // nothing to commit here.
+
+        debug("Commit: Nothing to commit");
+
```

Comment:
Maybe add a commented out example of adding a principal, like you did in the logout() method?